src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs

```diff
@@ -315,14 +329,31 @@ public async Task<LSP.CodeAction> HandleRequestAsync(LSP.CodeAction codeAction,
                         }
 
                         var edits = textChanges.Select(tc => ProtocolConversions.TextChangeToTextEdit(tc, oldText)).ToArray();
-                        var documentIdentifier = new OptionalVersionedTextDocumentIdentifier { Uri = newTextDoc.GetURI() };
-                        textDocumentEdits.Add(new TextDocumentEdit { TextDocument = documentIdentifier, Edits = edits });
+
+                        if (edits.Length > 0)
+                        {
+                            var documentIdentifier = new OptionalVersionedTextDocumentIdentifier { Uri = newTextDoc.GetURI() };
+                            textDocumentEdits.Add(new TextDocumentEdit { TextDocument = documentIdentifier, Edits = edits });
+                        }
+
+                        // Rename
+                        if (oldTextDoc.State.Attributes.Name != newTextDoc.State.Name)
+                        {
+                            textDocumentEdits.Add(new RenameFile() { OldUri = oldTextDoc.GetUriFromName(), NewUri = newTextDoc.GetUriFromName() });
+                        }
 
                         var linkedDocuments = solution.GetRelatedDocumentIds(docId);
                         modifiedDocumentIds.AddRange(linkedDocuments);
                     }
                 }
             }
         }
+
+        private static bool HasDocumentNameChange(DocumentId documentId, Solution newSolution, Solution oldSolution)
+        {
+            var newDocument = newSolution.GetRequiredDocument(documentId);
```

Comment:
This is going to throw for additional files, and should be calling GetRequiredTextDocument() instead.